pkg/river/internal/value/decode.go

```
@@ -0,0 +1,382 @@
+		into.Set(convVal.rv)
+		return nil
+	case TypeArray:
+		return decodeArray(convVal, into)
+	case TypeObject:
+		return decodeObject(convVal, into)
+	case TypeFunction:
+		// If the function types had the exact same signature, they would've been
+		// handled in the best case statement above. If we've hit this point,
+		// they're not the same.
+		//
+		// For now, we return an error.
+		//
+		// TODO(rfratto): we may want to consider being more lax here, potentially
+		// creating an adapter between the two functions.
+		return DecodeError{
+			Value: val,
+			Inner: fmt.Errorf("expected function(%s), got function(%s)", into.Type(), convVal.rv.Type()),
+		}
+	case TypeCapsule:
+		// Capsule types require being identical go types, which would've been
+		// handled in the best case statement above. If we hit this point, they're
+		// not the same.
+		//
+		// TODO(rfratto): return a TypeError for this instead. TypeError isn't
+		// appropriate at the moment because it would just print "capsule", which
+		// doesn't contain all the information the user would want to know (e.g., a
+		// capsule of what inner type?).
+		return DecodeError{
+			Value: val,
+			Inner: fmt.Errorf("expected capsule(%s), got %s", into.Type(), convVal.Describe()),
+		}
+	default:
+		panic("river/value: unexpected kind " + convVal.Type().String())
+	}
+}
+
+func decodeArray(val Value, rt reflect.Value) error {
+	switch rt.Kind() {
+	case reflect.Slice:
+		res := reflect.MakeSlice(rt.Type(), val.rv.Len(), val.rv.Len())
+		for i := 0; i < val.rv.Len(); i++ {
+			// Decode the original elements into the new elements.
+			if err := decode(val.Index(i), res.Index(i)); err != nil {
+				return ElementError{Value: val, Index: i, Inner: err}
+			}
+		}
+		rt.Set(res)
+
+	case reflect.Array:
+		res := reflect.New(rt.Type()).Elem()
+		for i := 0; i < val.rv.Len(); i++ {
+			// Stop processing elements if the target array is too short.
+			// TODO(rfratto): should we force array length to be identical?
```

Better than silently dropping elements IMHO. I think if someone's going through the trouble of using a Go array, they'd have a reason for wanting exactly N elements.

---

Yeah, I agree. Silently dropping is the behavior encoding/json uses, but I find it weird. I'll change it to enforce the length.